qtacrylicmainwindow.cpp

```diff
@@ -0,0 +1,199 @@
+    if (color.isValid() && (color != Qt::transparent)) {
+        return color;
+    } else {
+        return palette().color(backgroundRole());
+    }
+}
+
+void QtAcrylicMainWindow::setTintColor(const QColor &value)
+{
+    if (!value.isValid()) {
+        qWarning() << "Tint color not valid.";
+        return;
+    }
+    if (m_acrylicHelper.getTintColor() != value) {
+        m_acrylicHelper.setTintColor(value);
+        QPalette pal = palette();
+        pal.setColor(backgroundRole(), m_acrylicHelper.getTintColor());
+        setPalette(pal);
+        //m_acrylicHelper.updateAcrylicBrush(tintColor());
+        update();
+        Q_EMIT tintColorChanged();
+    }
+}
+
+qreal QtAcrylicMainWindow::tintOpacity() const
+{
+    return m_acrylicHelper.getTintOpacity();
+}
+
+void QtAcrylicMainWindow::setTintOpacity(const qreal value)
+{
+    if (m_acrylicHelper.getTintOpacity() != value) {
+        m_acrylicHelper.setTintOpacity(value);
+        m_acrylicHelper.updateAcrylicBrush(tintColor());
+        update();
+        Q_EMIT tintOpacityChanged();
+    }
+}
+
+qreal QtAcrylicMainWindow::noiseOpacity() const
+{
+    return m_acrylicHelper.getNoiseOpacity();
+}
+
+void QtAcrylicMainWindow::setNoiseOpacity(const qreal value)
+{
+    if (m_acrylicHelper.getNoiseOpacity() != value) {
+        m_acrylicHelper.setNoiseOpacity(value);
+        m_acrylicHelper.updateAcrylicBrush(tintColor());
+        update();
+        Q_EMIT noiseOpacityChanged();
+    }
+}
+
+bool QtAcrylicMainWindow::frameVisible() const
+{
+    return m_frameVisible;
+}
+
+void QtAcrylicMainWindow::setFrameVisible(const bool value)
+{
+    if (m_frameVisible != value) {
+        m_frameVisible = value;
+        update();
+        Q_EMIT frameVisibleChanged();
+    }
+}
+
+QColor QtAcrylicMainWindow::frameColor() const
+{
+    return m_acrylicHelper.getFrameColor();
+}
+
+void QtAcrylicMainWindow::setFrameColor(const QColor &value)
+{
+    if (m_acrylicHelper.getFrameColor() != value) {
+        m_acrylicHelper.setFrameColor(value);
+        update();
+        Q_EMIT frameColorChanged();
+    }
+}
+
+qreal QtAcrylicMainWindow::frameThickness() const
+{
+    return m_acrylicHelper.getFrameThickness();
+}
+
+void QtAcrylicMainWindow::setFrameThickness(const qreal value)
+{
+    if (m_acrylicHelper.getFrameThickness() != value) {
+        m_acrylicHelper.setFrameThickness(value);
+        update();
+        Q_EMIT frameThicknessChanged();
+    }
+}
+
+bool QtAcrylicMainWindow::acrylicEnabled() const
+{
+    return m_acrylicEnabled;
+}
+
+void QtAcrylicMainWindow::setAcrylicEnabled(const bool value)
+{
+    if (m_acrylicEnabled != value) {
+        m_acrylicEnabled = value;
+        if (m_acrylicEnabled) {
+            connect(&m_acrylicHelper, &QtAcrylicEffectHelper::needsRepaint, this, qOverload<>(&QtAcrylicMainWindow::update));
+        } else {        
+            disconnect(&m_acrylicHelper, &QtAcrylicEffectHelper::needsRepaint, this, qOverload<>(&QtAcrylicMainWindow::update));
```

Comment:
No need to disconnect, because the window frame is also draw by the acrylic helper. This signal is used by both the acrylic background and the window frame.

Reply:
And due to we will never disconnect, you should not make the connection here. Maybe move it to the showEvent is more appropriate.

Reply:
the frame is drawn in the `paintEvent()` why would you need extra repaints?

Reply:
the repaint signal is emited when the window state and activation changes, we need to re-draw the window frame in that case because the frame color need to be changed.

Reply:
the paintEvent won't be triggered if the window state or activation changes (because Qt think the window doesn't need to repaint), we need to trigger it manually.